rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,192 @@
+- Needed for Vector to act as a complete Datadog Agent aggregator (but further work will still be required).
+- Extend the Vector ecosystem, bring additional feature for distribution metrics that would enable consistent
+  aggregation.
+
+## Drawbacks
+
+- It arbitrary imposes internal sketches representation to follow the DDSketch paper.
+
+## Prior Art
+
+- Existing Vector metrics source over HTTP(s) works well, those constructs will be reused
+- Datadog logs is already supported in the `datadog_agent` source
+- Vector has its own vector-to-vector protocol that serve a similar purpose but the idea is to enable Vector to receive
+  Datadog metrics from unmodified Datadog Agents so that is probably not an option
+- There are many [ways](https://edoliberty.github.io/papers/streamingQuantiles.pdf) of implementing quantile sketches
+- Opentelemetry is moving to official support sketches on a [general
+  basis](https://github.com/open-telemetry/opentelemetry-specification/issues/982).
+
+## Alternatives
+
+- Use an alternate protocol between Datadog Agents and Vector (Like Prometheus, Statds or Vector own protocol)
+- For sketches, we could flatten sketches and compute usual derived metrics (min/max/average/count/some percentile) and
+  send those as gauge/count, but it would prevent (or at least impact) existing distribution/sketches user
+- Regarding how we support route and the fact that Vector would have to support non-metric payload, an alternate
+  solution would be to push for a Datadog Agent change with a new override (let's say `dd_metrics_url` that would only
+  divert request to `/api/v1/series` & `/api/beta/sketches` to a specific endpoints. This however highlights that Vector
+  & the Datadog Agent will then need to follow a compatbility matrix:
+  1. If the Agent is upgrade with a new metric route (let's say for v2 intake migration), the user will need to also
+     upgrade Vector so it supports this route.
+  2. Existing fleet of Agents will need to be upgraded to be able to send metrics to Vector.
+
+  While #2 is not really a major issue as this would be a one shot upgrade if required, #1 may be more bothersome as the
+  issue will always be there. A somehow hacky solution would be to leverage the [documented
+  haproxy](https://docs.datadoghq.com/agent/proxy/?tab=agentv6v7#haproxy) setup for Agent to divert selected routes to
+  Vector, but it would have the advantage of resolving any migrations, not-yet-supported-metric-route in Vector and
+  alleviate the need of modifying the Agent.
+
+## Outstanding Questions
+
+- Datadog Agent configuration, diverting metrics to a custom endpoints will likely divert other kind of traffic, this
+  need to be clearly identified / discussed
+- Is there any other metrics type that could required some degree of adaptation, there is the non monotonic counter that
+  exists in the agent but not in Vector that may require the introduction of a so-called `NonMonotonicCounter` in
+  Vector.
```

Comment:
This is a good point. We should discuss in this RFC how we'll handle drift between the two tools (i.e., forward and backward compatibility)

Reply:
FWIW, I believe we have this covered with the "absolute" mode for counters:https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L105

Reply:
Indeed, it could work, do you know if we already use "absolute" counter, because the [documentation](https://vector.dev/docs/about/under-the-hood/architecture/data-model/metric/#counter) is pretty authoritative and says "A single value that can be incremented or reset to a zero value but not decremented."